hashicorp/dev-portal
# feat: add signup form

> **Note**: 🏗 This PR targets the Certifications assembly branch. See #1447 for details.

## 🗒️ What

This PR implements the newsletter signup form. This form is shown at the bottom of all certifications pages.

## 🛠️ How

> **Note**: This PR uses a pre-release of `@hashicorp/react-marketo-form`, which is expected to be released before Certifications will launch. Once available, we'll open a PR onto the certifications assembly branch to bump to a stable release. See https://github.com/hashicorp/react-components/pull/816 for details.

- Plucks out existing signup form work from dot-com, adding styles
    - Props to Kendall for driving this - this PR cherry picks work done in #1431 during the certifications spike
- Adds mobile styles for the signup form
    - This is also cherry-picked from the certifications spike, from b4255d22
- Adds the signup form to program pages
    - During the spike, the signup form only appeared on the landing page. Designs have since been updated.
- Adds an `/api/marketo` route, using `@hashicorp/react-marketo-form`
    - This implementation is plucked from dot-com

> **Note**: I've also added `MARKETO` env variables to the Dev Dot project in Vercel - see [Dylan's comment below](https://github.com/hashicorp/dev-portal/pull/1469#issuecomment-1353793157) for details. These are necessary for the form submission process to function.

## 📸 Design Screenshots

🎨 [Figma mockups](https://www.figma.com/file/VD7ahvXuXWJApeGnhbW4hv/Developer?node-id=16598%3A247078&t=5WoCCWIPOOQqiG57-4)

### Design

| Mobile | Tablet | Desktop |
| - | - | - |
| <img width="1092" alt="design-responsive-guidance" src="https://user-images.githubusercontent.com/4624598/207977133-536bf42e-ca21-4e12-9566-22cddd1ba83e.png"> | <img width="1114" alt="design-tablet" src="https://user-images.githubusercontent.com/4624598/207977135-6e1c1727-f727-4ffd-93d5-2854da3914d1.png"> | <img width="1992" alt="design-desktop" src="https://user-images.githubusercontent.com/4624598/207977130-9b5cc20b-c2d0-4567-8444-75f6e27121b5.png"> |

### Implementation

| Mobile | Tablet | Desktop |
| - | - | - |
| <img width="377" alt="mobile-revised" src="https://user-images.githubusercontent.com/4624598/207991082-f1831c8b-57e6-4834-a463-a218f226e0dd.png"> | <img width="1062" alt="tablet-revised" src="https://user-images.githubusercontent.com/4624598/207991084-4555d990-38f8-4f8b-85c0-883cf58ad73d.png"> | <img width="1313" alt="desktop-revised" src="https://user-images.githubusercontent.com/4624598/207991085-5c55de51-fdf7-4af5-abbf-7ef25be05661.png"> | 

#### Form States

> **Note**: designs did not specify form states, I think because the assumption is they'd be implied from shared HDS input components. However, we have not yet implement shared HDS input components in React. This PR makes an effort to style all form states in a reasonable way. Intent is to handle further changes and refinement in VQA.

| State | Mobile | Desktop |
| - | - | - |
| Invalid Fields | <img width="378" alt="invalid-fields-mobile" src="https://user-images.githubusercontent.com/4624598/207991210-d0af04fa-3880-4343-a255-fe9d7fe41276.png"> | <img width="1256" alt="invalid-fields" src="https://user-images.githubusercontent.com/4624598/207991216-274a0d92-c265-44d5-b41f-aa8424753eb4.png"> |
| Failed submit | <img width="393" alt="failure-mobile" src="https://user-images.githubusercontent.com/4624598/207991202-a5dab4c6-7c20-40f4-b4db-81452ae27d25.png"> | <img width="1307" alt="failure" src="https://user-images.githubusercontent.com/4624598/207991213-cdb09d4b-bc41-48df-8350-d7a8a7365fdd.png"> |
| Successful submit | <img width="373" alt="success-mobile" src="https://user-images.githubusercontent.com/4624598/207991207-435cd27d-34c3-4d8c-a2a0-8fb0b5579890.png"> | <img width="1248" alt="success" src="https://user-images.githubusercontent.com/4624598/207991212-2e14741c-6edb-468c-b1db-a179f8268a22.png"> |

## 🔗 Preview links

- [/certifications][/certifications]
- [/certifications/infrastructure-automation][/certifications/infrastructure-automation]
- [/certifications/networking-automation][/certifications/networking-automation]
- [/certifications/security-automation][/certifications/security-automation]

## 🧪 Testing

- [ ] Visit each of the preview links, and confirm the signup form looks & functions as expected
    - The signup form should appear at the bottom of each page, above the footer
    - The signup form should be styled to match Dev Dot (see designs)
    - Submitting the signup form should result in a success message. (Note: submitting with an `@hashicorp.com` address is recommended, as those are filtered out in some way, as far as I understand at least).

To test the "failed submit" state:
- Pull this branch down locally
- Ignore the fact that you'd need `MARKETO` env variables to have the submission work (which you could get with `vercel env pull`, but don't do that)
- Try to submit the form. You should see an error message pop up, because `development`, but you should also see the form render the "failed" state.

[preview]: https://dev-portal-git-zscertifications-signup-hashicorp.vercel.app/
[/certifications]: https://dev-portal-git-zscertifications-signup-hashicorp.vercel.app/certifications
[/certifications/infrastructure-automation]: https://dev-portal-git-zscertifications-signup-hashicorp.vercel.app/certifications/infrastructure-automation
[/certifications/networking-automation]: https://dev-portal-git-zscertifications-signup-hashicorp.vercel.app/certifications/networking-automation
[/certifications/security-automation]: https://dev-portal-git-zscertifications-signup-hashicorp.vercel.app/certifications/security-automation
